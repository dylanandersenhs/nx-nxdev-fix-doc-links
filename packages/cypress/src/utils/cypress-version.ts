let cypressPackageJson;
let loadedCypress = false;

export function installedCypressVersion() {
  if (!loadedCypress) {
    try {
      cypressPackageJson = require('cypress/package.json');
    } catch {}
  }

  if (!cypressPackageJson) {
    return null;
  }
  const cypressPackageVersion = cypressPackageJson.version;
  const majorVersion = cypressPackageVersion.split('.')[0];
  if (!majorVersion) {
    return 0;
  }
  return +majorVersion;
}

export function assertCypress10Installed() {
  const version = installedCypressVersion();
  if (version && version < 10) {
    throw new Error(
      `Cypress version of 10 or higher is not installed. Expected Cypress v10+, found Cypress v${version} instead.`
    );
  }
}
