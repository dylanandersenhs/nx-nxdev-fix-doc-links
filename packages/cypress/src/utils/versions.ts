export const nxVersion = require('../../package.json').version;
export const eslintPluginCypressVersion = '^2.10.3';
export const typesNodeVersion = '16.11.7';
/**
 * TODO(caleb): this is only the mac version, and temp until there is a v10 release
 */
export const cypressVersion =
  'https://cdn.cypress.io/beta/npm/10.0.0/darwin-x64/10.0-release-ae078c74959ddd542ba6ad706613bfec1f07f761/cypress.tgz';
export const cypressWebpackVersion = '^1.8.0';
export const webpackHttpPluginVersion = '^5.5.0';
