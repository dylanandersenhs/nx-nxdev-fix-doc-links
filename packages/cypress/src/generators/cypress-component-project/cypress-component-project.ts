import {
  addDependenciesToPackageJson,
  formatFiles,
  generateFiles,
  joinPathFragments,
  offsetFromRoot,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { installedCypressVersion } from '../../utils/cypress-version';

import { CYPRESS_COMPONENT_TEST_TARGET } from '../../utils/project-name';
import {
  cypressVersion,
  cypressWebpackVersion,
  webpackHttpPluginVersion,
} from '../../utils/versions';
import { CypressComponentProjectSchema } from './schema';

export async function cypressComponentProject(
  tree: Tree,
  options: CypressComponentProjectSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const cyVersion = installedCypressVersion();

  if (cyVersion && cyVersion < 10) {
    throw new Error(
      'Cypress version of 10 or higher is required to create a component testing project'
    );
  }

  const installDepsTask = updateDeps(tree);

  addProjectFiles(tree, projectConfig, options);
  addTargetToProject(tree, projectConfig, options);

  if (!options.skipFormat) {
    await formatFiles(tree);
  }
  return () => {
    installDepsTask();
  };
}

function updateDeps(tree: Tree) {
  const devDeps = {
    '@cypress/webpack-dev-server': cypressWebpackVersion,
    'html-webpack-plugin': webpackHttpPluginVersion,
    cypress: cypressVersion,
  };

  return addDependenciesToPackageJson(tree, {}, devDeps);
}

function addProjectFiles(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  options: CypressComponentProjectSchema
) {
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    projectConfig.root,
    {
      ...options,
      projectRoot: projectConfig.root,
      offsetFromRoot: offsetFromRoot(projectConfig.root),
      ext: '',
    }
  );
}

function addTargetToProject(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  options: CypressComponentProjectSchema
) {
  projectConfig.targets[CYPRESS_COMPONENT_TEST_TARGET] = {
    executor: '@nrwl/cypress:cypress',
    options: {
      cypressConfig: joinPathFragments(projectConfig.root, 'cypress.config.ts'),
      testingType: 'component',
    },
  };

  updateProjectConfiguration(tree, options.project, projectConfig);
}
