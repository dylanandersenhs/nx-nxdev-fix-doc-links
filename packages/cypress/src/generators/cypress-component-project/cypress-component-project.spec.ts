import {
  addProjectConfiguration,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { CYPRESS_COMPONENT_TEST_TARGET } from '../../utils/project-name';

import { cypressComponentProject } from './cypress-component-project';

let projectConfig: ProjectConfiguration = {
  projectType: 'library',
  sourceRoot: 'libs/cool-lib/src',
  root: 'libs/cool-lib',
  targets: {
    build: {
      executor: '@nrwl/web:rollup',
      options: {
        tsConfig: 'libs/cool-lib/tsconfig.lib.json',
      },
    },
    test: {
      executor: '@nrwl/jest:jest',
      options: {
        jestConfig: 'libs/cool-lib/jest.config.js',
      },
    },
  },
};
describe('Cypress Component Project', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    addProjectConfiguration(tree, 'cool-lib', projectConfig);
    tree.write(
      '.gitignore',
      `
# compiled output
/dist
/tmp
/out-tsc
    `
    );
    tree.write(
      'libs/cool-lib/tsconfig.lib.json',
      `
{
  "extends": "./tsconfig.json",
  "exclude": [
    "**/*.spec.ts",
    "**/*.test.ts",
    "**/*.spec.tsx",
    "**/*.test.tsx",
    "**/*.spec.js",
    "**/*.test.js",
    "**/*.spec.jsx",
    "**/*.test.jsx",
    "**/*.cy.ts"
  ],
  "include": [
    "**/*.js",
    "**/*.jsx",
    "**/*.ts",
    "**/*.tsx"
  ]
}
`
    );
    tree.write(
      'libs/cool-lib/tsconfig.json',
      `
{
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ]
}
`
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should not error when rerunning on an existing project', async () => {
    tree.write('libs/cool-lib/cypress.config.ts', '');
    const newTarget = {
      [CYPRESS_COMPONENT_TEST_TARGET]: {
        executor: '@nrwl/cypress:cypress',
        options: {
          cypressConfig: 'libs/cool-lib/cypress.config.ts',
          testingType: 'component',
        },
      },
    };
    updateProjectConfiguration(tree, 'cool-lib', {
      ...projectConfig,
      targets: {
        ...projectConfig.targets,
        ...newTarget,
      },
    });

    await cypressComponentProject(tree, {
      project: 'cool-lib',
      skipFormat: true,
    });
    const actualProjectConfig = readProjectConfiguration(tree, 'cool-lib');

    expect(tree.exists('libs/cool-lib/cypress.config.ts')).toEqual(true);
    expect(tree.exists('libs/cool-lib/cypress')).toEqual(true);
    expect(tree.exists('libs/cool-lib/tsconfig.cy.json')).toEqual(true);
    expect(
      actualProjectConfig.targets[CYPRESS_COMPONENT_TEST_TARGET]
    ).toMatchSnapshot();
  });
});
