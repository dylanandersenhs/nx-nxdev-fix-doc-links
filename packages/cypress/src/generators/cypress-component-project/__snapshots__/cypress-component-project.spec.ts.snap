// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cypress Component Project should add base cypress component testing config 1`] = `
Object {
  "executor": "@nrwl/cypress:cypress",
  "options": Object {
    "cypressConfig": "libs/cool-lib/cypress.config.ts",
    "testingType": "component",
  },
}
`;

exports[`Cypress Component Project should not error when rerunning on an existing project 1`] = `
Object {
  "executor": "@nrwl/cypress:cypress",
  "options": Object {
    "cypressConfig": "libs/cool-lib/cypress.config.ts",
    "testingType": "component",
  },
}
`;
